usdigitalresponse/grants-ingest
# 2/split grants gov xml

### Ticket #2 

## Description

This PR adds a new `SplitGrantsGovXMLDB` Lambda function per the deliverable specified in #2. It includes both runtime source code as well as Terraform for provisioning the Lambda function and its required resources.

A few things that may be useful to call out explicitly:
1. Since this Lambda function makes use of a goroutine-based worker pool to concurrently process individual parsed records (any of which could potentially cause an error), as well as the fact that we don't allow a single bad record to cause the entire Lambda function to fail, the handler code makes use of the [`hashicorp/go-multierror` library](https://github.com/hashicorp/go-multierror) in order to aggregate errors, which is used to signal Lambda invocation failure/success state. Therefore, if at least one error was encountered during processing, the invocation will appear as failed, even when many/most/all other processing operations succeed. Additionally, since the handler code considers a source S3 object in its entirety, retrying after a failed invocation results in everything being re-processed. 
    - This seems like an ideal starting point, and may work out just fine in the long run.
    - Some  kind of "bisect-on-error" or record-based retry behavior may be warranted if we notice that per-record failures occur often enough to be a hassle.
    - Each processed record incurs exactly one `HeadObject` call to S3, and up to one `PutObject` call. During a retry, records that successfully processed during the previous invocations for the same source object will not incur in a `PutObject` call, but will always incur a `HeadObject` call. These `HeadObject` requests are free in terms of S3 cost (but not Lambda runtime cost).
2. Currently, there are >70k individual grant opportunity records parsed from the source XML database dump. As noted previously, each invocation makes 1 or 2 calls to S3 per source record parsed from the XML DB dump, meaning that each invocation currently makes between 70k - 140k S3 API calls.  
    - Since we automatically instrument tracing on all AWS SDK calls with Datadog, this produces a similar number of spans within the invocation, which currently seems to cause the Datadog agent's trace collector to drop the trace, presumably(? – I think it's odd that the DD library would fail here because it's not _so_ much data in the grand scheme of things) because the trace payload grows too large.
    - I have an open ticket with Datadog support regarding this issue, but it's currently in the "anticipate a response from engineering between 1-3, but maybe more depending on backlog" phase and it's been 4 days since my last update 🤷 
    - I don't think it's necessary to hold this PR up until this gets resolved, but presumably we'll need to follow up at some point once we get some advice on how to appropriately tune the agent settings.
3. This PR also adds a new package under `pkg/grantsSchemas/grants.gov`. Aside from a few helper functions that turn date/time strings into native Go `time.Time` values and some light aliasing, the Go code in this package is dynamically generated (using `go generate`) from the [`OpportunityDetail-V1.0` XML schema](https://apply07.grants.gov/apply/system/schemas/OpportunityDetail-V1.0.xsd) provided by Grants.gov. Generated contents, including Go code and the XML schema file itself, are meant to be committed. They can be refreshed by running `task generate`.

## Changes to Metrics and Monitoring

- This Lambda emits 4 different distribution metrics (all namespaced with `grants_ingest.split_grants_gov_xml_db`):
  - `grants_ingest.split_grants_gov_xml_db.opportunity.created`: number of grant records saved to S3 that did not overwrite an existing object.
  - `grants_ingest.split_grants_gov_xml_db.opportunity.updated`: number of grant opportunities saved to S3 that did overwrite (i.e. replaced) an existing object.
  - `grants_ingest.split_grants_gov_xml_db.opportunity.skipped`: number of grant opportunities not saved to S3 because the existing S3 data was already up-to-date.
  - `grants_ingest.split_grants_gov_xml_db.opportunity.failed`: number of grant opportunities not saved to S3 because an error was encountered during processing.
- The sum of all four of these metrics within a window should equal the number of grant opportunities read from the XML source file(s) over the same window.
- All emitted metrics are tagged with `source:grants.gov`.

## Testing

To test using LocalStack:
1. With this branch checked out, prepare LocalStack [as documented in our README](https://github.com/usdigitalresponse/grants-ingest/#provisioning-infrastructure).
2. Prepare a source XML file that conforms to the Grants.gov `OpportunityDetail-V1.0` schema. To obtain a real one, follow these steps:
    1. Go to https://www.grants.gov/xml-extract.html and download an available zip file.
    2. Unzip the downloaded file, which contains a single file named according to the date of the DB dump, e.g. `GrantsDBExtract20230323v2.xml`.
3. Upload the XML file to the source S3 bucket using the appropriate S3 key. For the example file named previously in Step 2.2, this would be `sources/2023/03/23/grants.gov/extract.xml`. Example command using `awslocal`:
    ```cli
    $ awslocal s3 cp --sse AES256 ./GrantsDBExtract20230323v2.xml s3://grants-ingest-grantssourcedata-000000000000-us-west-2/sources/2023/03/23/grants.gov/extract.xml
    ```
4. Once the upload finishes, ensure that the Lambda invocation was triggered.
5. Wait for the invocation to finish (this may take a while depending on system resources and source file size) and verify that the invocation completed successfully. You can also check the logs for any errors.
6. Verify that new objects exist in the destination bucket. You can use this `awslocal` command to recursively list the contents of the S3 bucket, but be aware that if you used a real Grants.gov DB dump for testing, the output will be quite large:
    ```cli
    $ awslocal s3 ls --recursive s3://grants-ingest-grantssourcedata-000000000000-us-west-2
    ```

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
